File: src/transformers/features/json/JsonApplicationTransformer.ts
// import ts from "typescript";

// import { LiteralFactory } from "../../../factories/LiteralFactory";
// import { MetadataCollection } from "../../../factories/MetadataCollection";
// import { MetadataFactory } from "../../../factories/MetadataFactory";

// import { IJsonApplication } from "../../../schemas/json/IJsonApplication";
// import { Metadata } from "../../../schemas/metadata/Metadata";

// import { JsonApplicationProgrammer } from "../../../programmers/json/JsonApplicationProgrammer";

// import { ValidationPipe } from "../../../typings/ValidationPipe";

// import { ITransformProps } from "../../ITransformProps";
// import { TransformerError } from "../../TransformerError";

// export namespace JsonApplicationTransformer {
//   export const transform = (props: ITransformProps): ts.Expression => {
//     // GET GENERIC ARGUMENT
//     if (!props.expression.typeArguments?.length)
//       throw new TransformerError({
//         code: "typia.json.application",
//         message: "no generic argument.",
//       });

//     const top: ts.Node = props.expression.typeArguments[0]!;
//     if (ts.isTypeNode(top) === false) return props.expression;

//     const version: "3.0" | "3.1" = get_parameter<"3.0" | "3.1">({
//       checker: props.context.checker,
//       name: "Version",
//       is: (str) => str === "3.0" || str === "3.1",
//       cast: (str) => str as "3.0" | "3.1",
//       default: () => "3.1",
//     })(props.expression.typeArguments[1]);

//     // GET TYPE
//     const type: ts.Type = props.context.checker.getTypeFromTypeNode(top);
//     const collection: MetadataCollection = new MetadataCollection({
//       replace: MetadataCollection.replace,
//     });
//     const result: ValidationPipe<Metadata, MetadataFactory.IError> =
//       MetadataFactory.analyze({
//         checker: props.context.checker,
//         transformer: props.context.transformer,
//         options: {
//           escape: true,
//           constant: true,
//           absorb: false,
//           functional: true,
//           validate: JsonApplicationProgrammer.validate,
//         },
//         collection,
//         type,
//       });
//     if (result.success === false)
//       throw TransformerError.from({
//         code: "typia.json.application",
//         errors: result.errors,
//       });

//     // GENERATE LLM APPLICATION
//     const app: IJsonApplication<"3.0" | "3.1"> =
//       JsonApplicationProgrammer.write({
//         version,
//         metadata: result.data,
//       });
//     const literal: ts.Expression = LiteralFactory.write(app);
//     return literal;
//   };

//   const get_parameter =
//     <Value>(props: {
//       checker: ts.TypeChecker;
//       name: string;
//       is: (value: string) => boolean;
//       cast: (value: string) => Value;
//       default: () => Value;
//     }) =>
//     (node: ts.TypeNode | undefined): Value => {
//       if (!node) return props.default();

//       // CHECK LITERAL TYPE
//       const type: ts.Type = props.checker.getTypeFromTypeNode(node);
//       if (
//         !type.isLiteral() &&
//         (type.getFlags() & ts.TypeFlags.BooleanLiteral) === 0
//       )
//         throw new TransformerError({
//           code: "typia.json.application",
//           message: `generic argument "${props.name}" must be constant.`,
//         });

//       // GET VALUE AND VALIDATE IT
//       const value = type.isLiteral()
//         ? type.value
//         : props.checker.typeToString(type);
//       if (typeof value !== "string" || props.is(value) === false)
//         throw new TransformerError({
//           code: "typia.json.application",
//           message: `invalid value on generic argument "${props.name}".`,
//         });
//       return props.cast(value);
//     };
// }

File: packages/typia/native/transform/features/json/JsonApplicationTransformer.go
package json

import (
  shimast "github.com/microsoft/typescript-go/shim/ast"
  nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
  nativejsonprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers/json"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
  nativetransform "github.com/samchon/typia/packages/typia/native/transform/internal"
)

type jsonApplicationTransformerNamespace struct{}

var JsonApplicationTransformer = jsonApplicationTransformerNamespace{}

func (jsonApplicationTransformerNamespace) Transform(props nativetransform.ITransformProps) *shimast.Node {
  if props.Expression == nil || props.Expression.TypeArguments == nil || len(props.Expression.TypeArguments.Nodes) == 0 {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
      Code:    "typia.json.application",
      Message: "no generic argument.",
    }))
  }

  top := props.Expression.TypeArguments.Nodes[0]
  if top == nil {
    return props.Expression.AsNode()
  }

  var versionNode *shimast.Node
  if len(props.Expression.TypeArguments.Nodes) > 1 {
    versionNode = props.Expression.TypeArguments.Nodes[1]
  }
  version := jsonTransformer_getParameter(jsonTransformer_getParameterProps{
    Checker: props.Context.Checker,
    Code:    "typia.json.application",
    Name:    "Version",
    Is: func(str string) bool {
      return str == "3.0" || str == "3.1"
    },
    Default: func() string {
      return "3.1"
    },
  })(versionNode)

  typ := props.Context.Checker.GetTypeFromTypeNode(top)
  collection := schemametadata.NewMetadataCollection(&schemametadata.MetadataCollection_IOptions{
    Replace: schemametadata.MetadataCollection_replace,
  })
  result := nativefactories.MetadataFactory.Analyze(nativefactories.MetadataFactory_IProps{
    Checker:     props.Context.Checker,
    Transformer: props.Context.Transformer,
    Options: nativefactories.MetadataFactory_IOptions{
      Escape:     true,
      Constant:   true,
      Absorb:     false,
      Functional: true,
      Validate:   jsonTransformer_applicationValidator,
    },
    Components: collection,
    Type:       typ,
  })
  if result.Success == false {
    panic(nativetransform.TransformerError_from(struct {
      Code   string
      Errors []nativetransform.TransformerError_MetadataFactory_IError
    }{
      Code:   "typia.json.application",
      Errors: jsonTransformer_errors(result.Errors),
    }))
  }

  return nativejsonprogrammers.JsonApplicationProgrammer.Write(nativejsonprogrammers.JsonApplicationProgrammer_IWriteProps{
    Context:  props.Context,
    Version:  version,
    Metadata: result.Data,
  })
}
